relay-server/src/http.rs

```diff
@@ -0,0 +1,299 @@
+///! objects and common request objects, it's just that nobody bothers to implement the conversion
+///! logic.
+use std::io;
+use std::io::Write;
+
+use actix_web::client::{ClientRequest, ClientRequestBuilder, ClientResponse};
+use actix_web::http::{ContentEncoding, StatusCode};
+use actix_web::{Binary, Error as ActixError, HttpMessage};
+use brotli2::write::BrotliEncoder;
+use flate2::write::{GzEncoder, ZlibEncoder};
+use flate2::Compression;
+use futures::{future, prelude::*};
+use futures03::{FutureExt, TryFutureExt};
+use serde::de::DeserializeOwned;
+
+use ::actix::prelude::*;
+
+use relay_config::HttpEncoding;
+
+pub enum Request {
+    Actix(ClientRequest),
+    Reqwest(reqwest::Request),
+}
+
+pub enum RequestBuilder {
+    Actix(ClientRequestBuilder),
+    Reqwest {
+        builder: reqwest::RequestBuilder,
+
+        /// The content encoding that this builder object implements on top of reqwest, which does
+        /// not support request encoding at all.
+        http_encoding: HttpEncoding,
+    },
+}
+
+impl RequestBuilder {
+    pub fn reqwest(builder: reqwest::RequestBuilder) -> Self {
+        RequestBuilder::Reqwest {
+            builder,
+
+            // very few endpoints can actually deal with request body content-encoding. Outside of
+            // store/envelope this is almost always identity because there's no way to do content
+            // negotiation on request.
+            http_encoding: HttpEncoding::Identity,
+        }
+    }
+
+    pub fn actix(builder: ClientRequestBuilder) -> Self {
+        RequestBuilder::Actix(builder)
+    }
+
+    pub fn no_default_headers(&mut self) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.no_default_headers();
+            }
+            RequestBuilder::Reqwest { .. } => {
+                // can only be set on client
+            }
+        }
+
+        self
+    }
+
+    pub fn finish<E>(self) -> Result<Request, E>
+    where
+        E: From<reqwest::Error> + From<ActixError>,
+    {
+        match self {
+            RequestBuilder::Actix(mut builder) => Ok(Request::Actix(builder.finish()?)),
+            RequestBuilder::Reqwest {
+                builder,
+                http_encoding: _,
+            } => Ok(Request::Reqwest(builder.build()?)),
+        }
+    }
+
+    /// Add a new header, not replacing existing ones.
+    pub fn header(&mut self, key: &str, value: &[u8]) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.header(key, value);
+            }
+            RequestBuilder::Reqwest { builder, .. } => {
+                take_mut::take(builder, |b| b.header(key, value))
+            }
+        }
+
+        self
+    }
+
+    pub fn body<E>(self, body: Binary) -> Result<Request, E>
+    where
+        E: From<reqwest::Error> + From<ActixError> + From<io::Error>,
```

Comment:
this looks a bit weird. Maybe its better to return a concrete error enum as well and have callers do conversion?

Reply:
I changed it but now you have to explicitly convert the new type in builders to the upstream error type